docs/Configuration.md
# Log Courier Configuration

<!-- START doctoc generated TOC please keep comment here to allow auto update -->
<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->
**Table of Contents**  *generated with [DocToc](https://github.com/thlorenz/doctoc)*

- [Overview](#overview)
- [YAML Format](#yaml-format)
- [JSON Format](#json-format)
- [Examples](#examples)
- [Reloading](#reloading)
- [Field Types](#field-types)
  - [String, Number, Boolean, Array, Dictionary](#string-number-boolean-array-dictionary)
  - [Duration](#duration)
  - [Fileglob](#fileglob)
- [Stream Configuration](#stream-configuration)
  - [`add host field`](#add-host-field)
  - [`add offset field`](#add-offset-field)
  - [`add path field`](#add-path-field)
  - [`add timezone field`](#add-timezone-field)
  - [`codecs`](#codecs)
  - [`dead time`](#dead-time)
  - [`unlinked dead time`](#unlinked-dead-time)
  - [`fields`](#fields)
- [`admin`](#admin)
  - [`enabled`](#enabled)
  - [`listen address`](#listen-address)
- [`files`](#files)
  - [`paths`](#paths)
- [`general`](#general)
  - [`log file`](#log-file)
  - [`global fields`](#global-fields)
  - [`host`](#host)
  - [`log level`](#log-level)
  - [`log stdout`](#log-stdout)
  - [`log syslog`](#log-syslog)
  - [`line buffer bytes`](#line-buffer-bytes)
  - [`max line bytes`](#max-line-bytes)
  - [`persist directory`](#persist-directory)
  - [`prospect interval`](#prospect-interval)
  - [`spool max bytes`](#spool-max-bytes)
  - [`spool size`](#spool-size)
  - [`spool timeout`](#spool-timeout)
  - [`listen address`](#listen-address)
  - [`telemetry path`](#telemetry-path)
- [`includes`](#includes)
- [`network`](#network)
  - [`failure backoff`](#failure-backoff)
  - [`failure backoff max`](#failure-backoff-max)
  - [`max pending payloads`](#max-pending-payloads)
  - [`method`](#method)
  - [`reconnect backoff`](#reconnect-backoff)
  - [`reconnect backoff max`](#reconnect-backoff-max)
  - [`rfc 2782 srv`](#rfc-2782-srv)
  - [`rfc 2782 service`](#rfc-2782-service)
  - [`servers`](#servers)
  - [`ssl ca`](#ssl-ca)
  - [`ssl certificate`](#ssl-certificate)
  - [`ssl key`](#ssl-key)
  - [`timeout`](#timeout)
  - [`transport`](#transport)
- [`stdin`](#stdin)

<!-- END doctoc generated TOC please keep comment here to allow auto update -->

## Overview

The Log Courier configuration file is in YAML format. It can also be in JSON
format is desired.

When loading a configuration file, the format is determined by the file
extension. If the file extension is not recognised, Log Courier reports an
error.

* `.yaml` - YAML
* `.json` - JSON
* `.conf` - JSON

## YAML Format

YAML format is the preferred format for Log Courier configuration files as it
tends to be easier to read and write than most other formats, such as JSON. It
is also easier to manage by scripts as it is line-based.

Entries start with their name followed by a colon. A value then follows it, with
optional quoting, on the same line, or a block of names and values follow it on
the next line with indentation.

Lists can be specified by starting each line with `-`. To place a block in a
list, prefix the first name and value pair with `-` and then indent the rest
accordingly.

It is best explained with an example:

```
network:
  servers:
    - "10.0.0.1:12345"
    - "10.0.0.2:12345"
  transport: tcp
files:
  - paths: [ "/var/log/nginx/access.log", "/var/log/nginx/error.log" ]
    fields:
      type: nginx_logs
```

## JSON Format

If in doubt, it is best to use YAML configuration as it tends to be easier to
read and write than JSON format.

Blocks are wrapped in `{` and `}` and within those is a comma-separate list of
name and value pairs. Names are always double-quoted strings
(e.g. `"hello world"`) and the values can be numbers, double-quoted strings,
another block, or a list of something. Lists are comma-separated lists of these
and are wrapped in `[` and `]`. The configuration file is formed by a
single block, as shown below.

```
{
  "network": {
    "servers": [ "10.0.0.1:12345", "10.0.0.2:12345" ],
    "transport": "tcp"
  },
  "files": [ {
    "paths": [ "/var/log/nginx/access.log", "/var/log/nginx/error.log" ]
  } ]
}
```

End-of-line comments start with a pound sign outside of a string, and cause all
characters until the end of the line to be ignored. Block comments start with
a forwarder slash and an asterisk and cause all characters, including new lines,
to be ignored until an asterisk followed by a forwarder slash is encountered.

```
{
    "general": {
        # General configuration here
    },
    "network": {
        # Network configuration here
    }, # (these are end-of-line comments)
    "files": [
        /* File configuration here
        (this is a block comment) */
    ]
}
```

## Examples

Several configuration examples are available for perusal in the
[examples folder](examples). Examples use the preferred format of YAML.

* [Ship a single log file](examples/example-single.yaml)
* [Ship a folder of log files](examples/example-folder.yaml)
* [Ship from STDIN](examples/example-stdin.yaml)
* [Ship logs with extra field information](examples/example-fields.yaml)
* [Multiline log processing](examples/example-multiline.yaml)
* [Multiple codecs](examples/example-multiple-codecs.yaml)

An example [JSON configuration](examples/example-json.conf) is also available
that follows a single log file.

The configuration is documented in full below.

## Reloading

Log Courier can reload its configuration without the need for a restart. It will
do this upon receiving the SIGHUP signal. To send this signal, run the following
command replacing 1234 with the Process ID of Log Courier.

    kill -HUP 1234

Log Courier will reopen its own log file if one has been configured, allowing
native log rotation to take place.

Please note that files Log Courier has already started harvesting will continue
to be harvested after the reload with their original configuration; the reload
process will only affect new files. Additionally, harvested log files will not
be reopened. Log rotations are detected automatically. To control when a
harvested log file is closed you can adjust the [`dead time`](#dead-time) and
[`unlinked dead time`](#unlinked-dead-time) options.

In the case of a network configuration change, Log Courier will disconnect and
reconnect as required at the earliest opportunity.

*Configuration reload is not currently available on Windows builds of Log
Courier.*

## Field Types

### String, Number, Boolean, Array, Dictionary

These are JSON types and follow the same rules. Strings within double quotes,
arrays of fields within square brackets separated by commas, and dictionaries
of key value pairs within curly braces and each entry, in the form `key":
value`, separated by a comma.

### Duration

This can be either a number or a string describing the duration. A number will
always be interpreted in seconds.

* `5` = 5 seconds
* `300` = 5 minutes (which is 300 seconds)
* `5s` = 5 seconds
* `15m` = 15 minutes

### Fileglob

A fileglob is a string representing a file pattern.

The pattern format used is detailed at
http://golang.org/pkg/path/filepath/#Match and is shown below for reference:

```
term:
    '*'         matches any sequence of non-Separator characters
    '?'         matches any single non-Separator character
    '[' [ '^' ] { character-range } ']'
                            character class (must be non-empty)
    c           matches character c (c != '*', '?', '\\', '[')
    '\\' c      matches character c

character-range:
    c           matches character c (c != '\\', '-', ']')
    '\\' c      matches character c
    lo '-' hi   matches character c for lo <= c <= hi
```

* `/var/log/*.log`
* `/var/log/program/log_????.log`
* `/var/log/httpd/access.log`
* `/var/log/httpd/access.log.[0-9]`

## Stream Configuration

Stream Configuration parameters can be specified for file groups within
[`files`](#files) and also for [`stdin`](#stdin). They customise the log
entries produced by passing, for example, by passing them through a codec and
adding extra fields.

### `add host field`

*Boolean. Optional. Default: true*

Adds an automatic "host" field to generated events that contains the `host`
value from the general configuration section.

### `add offset field`

*Boolean. Optional. Default: true*

Adds an automatic "offset" field to generated events that contains the current
offset in the current data stream.

*Beware that this value will reset when a file rotates or is truncated and is
generally not useful. It will be kept configurable to allow full compatibility
with Logstash Forwarder's traditional behaviour, and from version 2 the default
will be changed to false.*

### `add path field`

*Boolean. Optional. Default: true*

Adds an automatic "path" field to generated events that contains the path to the
current data stream. For stdin, this field is set to a hyphen, "-".

### `add timezone field`

*Boolean. Optional. Default: false*

Adds an automatic "timezone" field to generated events that contains the local
machine's local timezone in the format, "-0700 MST".

### `codecs`

*Codec configuration. Optional. Default: Single `plain` codec  
Configuration reload will only affect new or resumed files*

*Depending on how log-courier was built, some codecs may not be available. Run
`log-courier -list-supported` to see the list of codecs available in a specific
build of log-courier.*

The specified codecs will receive the lines read from the log stream and perform
any decoding necessary to generate events. The plain codec does nothing and
simply ships the events unchanged.

When multiple codecs are specified, the first codec will receive events, and the
second codec will receive the output from the first codec. This allows versatile
configurations, such as combining events into multiline events and then
filtering those that aren't required.

All configurations are an array of dictionaries with at least a "name" key.
Additional options can be provided if the specified codec allows.

* `[ { "name": "codec-name" } ]`
* `[ { "name": "codec-name", "option1": "value", "option2": "42" } ]`
* `[ { "name": "first-name" }, { "name": "second-name" } ]`

Aside from "plain", the following codecs are available at this time.

* [Filter](codecs/Filter.md)
* [Multiline](codecs/Multiline.md)

### `dead time`

*Duration. Optional. Default: "1h"  
Configuration reload will only affect new or resumed files*

If a log file has not been modified in this time period, it will be closed and
Log Courier will simply watch it for modifications. If the file is modified it
will be reopened.

### `unlinked dead time`

*Duration. Optional. Default: "1h"
Configuration reload will only affect new or resumed files*

If a log file that is being harvested is deleted, it will remain on disk until
Log Courier closes it. On Linux, Log Courier detects that a file has been
deleted, and will keep it open for the shorter of this and `dead time`, to
allow the space on disk to be released as soon as possible.

Don't reduce this unless you're sure that the log file rotation processes in
use guarantee there's no possibility of an application continuing to write to a
long-unlinked log file.

### `fields`

*Dictionary. Optional  
Configuration reload will only affect new or resumed files*

Extra fields to attach to events prior to shipping. These can be simple strings,
numbers or even arrays and dictionaries.

Examples:

* `{ "type": "syslog" }`
* `{ "type": "apache", "server_names": [ "example.com", "www.example.com" ] }`
* `{ "type": "program", "program": { "exec": "program.py", "args": [ "--run", "--daemon" ] } }`

## `admin`

The admin configuration enables or disabled the REST interface within Log
Courier and allows you to configure the interface to listen on.

### `enabled`

*Boolean. Optional. Default: false  
Requires restart*

Enables the REST interface. The `lc-admin` utility can be used to connect to
this.

### `listen address`

*String. Required when `enabled` is true.  
Default: tcp:127.0.0.1:1234  
RPM/DEB Package Default: unix:/var/run/log-courier/admin.socket*

The address the REST interface should listen on must be in the format
`transport:address`.

Allowed transports are "tcp", "tcp4", "tcp6" (Windows and *nix) and "unix"
(*nix only). For the tcp transports the address format is `host:port`. For the
unix transport the address should specify a filename to use when creating the
unix domain socket. If no transport name is specified, "tcp" is assumed.

Examples:

    127.0.0.1:1234
    tcp:127.0.0.1:1234
    unix:/var/run/log-courier/admin.socket

## `files`

The files configuration lists the file groups that contain the logs you wish to
ship. It is an array of file group configurations.

```
    [
        {
            # First file group
        },
        {
            # Second file group
        }
    ]
```

In addition to the configuration parameters specified below, each file group may
also have [Stream Configuration](#stream-configuration) parameters specified.

### `paths`

*Array of Fileglobs. Required*

At least one Fileglob must be specified and all matching files for all provided
globs will be monitored.

If the log file is rotated, Log Courier will detect this and automatically start
harvesting the new file. It will also keep the old file open to catch any
delayed writes that a still-reloading application has not yet written. You can
configure the time period before this old log file is closed using the
[`dead time`](#dead-time) and [`unlinked dead time`](#unlinked-dead-time)
options.

See above for a description of the Fileglob field type.

*To read from stdin, see the [`-stdin`](CommandLineArguments.md#stdin) command
line argument.*

Examples:

* `[ "/var/log/*.log" ]`
* `[ "/var/log/program/log_????.log" ]`
* `[ "/var/log/httpd/access.log", "/var/log/httpd/access.log.[0-9]" ]`

## `general`

The general configuration affects the general behaviour of Log Courier, such
as where to store its persistence data or how often to scan for the appearence
of new log files.

### `log file`

*Filepath. Optional  
Requires restart*

A log file to save Log Courier's internal log into. May be used in conjunction with `log stdout` and `log syslog`.

### `global fields`

*Dictionary. Optional  
Configuration reload will only affect new or resumed files*

Extra fields to attach to events prior to shipping. This is identical in
behaviour to the `fields` Stream Configuration and applies globally to the
`stdin` section and to all files listed in the `files` section.

### `host`

*String. Optional. Default: System FQDN.  
Configuration reload will only affect new or resumed files*

Every event has an automatic field, "host", that contains the current system
FQDN. Using this option allows a custom value to be given to the "host" field
instead of the system FQDN.

### `log level`

*String. Optional. Default: "info".  
Available values: "critical", "error", "warning", "notice", "info", "debug"  
Requires restart*

The minimum level of detail to produce in Log Courier's internal log.

### `log stdout`

*Boolean. Optional. Default: true  
Requires restart*

Enables sending of Log Courier's internal log to the console (stdout). May be used in conjunction with `log syslog` and `log file`.

### `log syslog`

*Boolean. Optional. Default: false  
Requires restart*

Enables sending of Log Courier's internal log to syslog. May be used in conjunction with `log stdout` and `log file`.

*This option is ignored by Windows builds.*

### `line buffer bytes`

*Number. Optional. Default: 16384*

The size of the line buffer used when reading files.

If `max line bytes` is greater than this value, any lines that exceed this size
will trigger additional memory allocations. This value should be set to a value
just above the 90th percentile (or average) line length.

### `listen address`

*String. Optional. Default: (disabled)*

If configured, log-courier will listen on HTTP and serve metrics that can be
collected by Prometheus at the [`telemetry path`](#telemetry-path).

    "listen address": ":10180"

By default, the listener is disabled.

### `max line bytes`

*Number. Optional. Default: 1048576*

The maxmimum line length to process. If a line exceeds this length, it will be
split across multiple events. Each split line will have a "tag" field added
containing the tag "splitline". The final part of the line will not have a "tag"
field added.

If the `fields` configuration already contained a "tags" entry, and it is an
array, it will be appended to. Otherwise, the "tag" field will be left as is.

This setting can not be greater than the `spool max bytes` setting.

### `persist directory`

*String. Required  
RPM/DEB Package Default: /var/lib/log-courier
Requires restart*

The directory that Log Courier should store its persistence data in.

At the time of writing, the only file saved here is `.log-courier` that
contains the offset in the file that Log Courier needs to resume from after a
graceful restart or crash. The offset is only updated when the remote endpoint
acknowledges receipt of the events.

### `prospect interval`

*Duration. Optional. Default: 10*

How often Log Courier should check for changes on the filesystem, such as the
appearance of new log files, rotations and deletions.

### `spool max bytes`

*Number. Optional. Default: 10485760*

The maximum size of an event spool, before compression. If an incomplete spool
does not have enough room for the next event, it will be flushed immediately.

If this value is modified, the receiving end should also be configured with the
new limit. For the Logstash plugin, this is the `max_packet_size` setting.

The maximum value for this setting is 2147483648 (2 GiB).

### `spool size`

*Number. Optional. Default: 1024*

How many events to spool together and flush at once. This improves efficiency
when processing large numbers of events by submitting them for processing in
bulk.

Internal benchmarks have shown that increasing to 5120, for example, can
give around a 25% boost of events per second, at the expense of more memory
usage.

*For most installations you should leave this at the default as it can
easily cope with over 10,000 events a second and uses little memory. It is
useful only in very specific circumstances.*

### `spool timeout`

*Duration. Optional. Default: 5*

The maximum amount of time to wait for a full spool. If an incomplete spool is
not filled within this time limit, the spool will be flushed immediately.

### `telemetry path`

*String. Optional. Default: `/metrics`*

The path Prometheus metrics are exposed at.

With a [`listen address`](#`listen-address`) of `:10180` and the default
`telemetry path`, the metrics can be accessed at
`http://(host ip):10180/metrics`.

## `includes`

*Array of Fileglobs. Optional*

Includes should be an array of additional file group configuration files to
read. Each configuration file should follow the format of the `files` section.

    "includes": [ "/etc/log-courier/conf.d/*.conf" ]

A file at `/etc/log-courier/conf.d/apache.conf` could then contain the
following.

    [ {
        "paths": [ "/var/log/httpd/access.log" ],
        "fields": { "type": "access_log" }
    } ]

## `network`

The network configuration tells Log Courier where to ship the logs, and also
what transport and security to use.

### `failure backoff`

*Duration. Optional. Default: 0*

Pause this long before allowing a failed endpoint to be used again. On each
consecutive failure, the pause is exponentially increased.

When set to 0, there is no backoff after the first failure. The second failure
then pauses for 1 second and begins to exponentially increase on each
consecutive failure.

(This is distinct from `reconnect backoff` in that it handles all failures, and
is available for all transports. For example, if a remote endpoint keeps failing
almost immediately after a successful connection, this backoff would prevent it
from being used again immediately and slowing down the entire pipeline.)

### `failure backoff max`

*Duration. Optional. Default: 300s*

The maximum time to wait before using a failed endpoint again. This prevents the
exponential increase of `failure backoff` from becoming too high.

### `max pending payloads`

*Number. Optional. Default: 4*


lc-lib/config/config.go
/*
 * Copyright 2014-2015 Jason Woods.
 *
 * This file is a modification of code from Logstash Forwarder.
 * Copyright 2012-2013 Jordan Sissel and contributors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package config

import (
	"fmt"
	"os"
	"path"
	"path/filepath"
	"reflect"
	"time"

	"github.com/driskell/log-courier/lc-lib/addresspool"
	"gopkg.in/op/go-logging.v1"
)

var (
	// DefaultConfigurationFile is a path to the default configuration file to
	// load, this can be changed during init()
	DefaultConfigurationFile = ""

	// DefaultGeneralPersistDir is a path to the default directory to store
	DefaultGeneralPersistDir = ""
)

const (
	defaultGeneralHost               string        = "localhost.localdomain"
	defaultGeneralLogLevel           logging.Level = logging.INFO
	defaultGeneralLogStdout          bool          = true
	defaultGeneralLogSyslog          bool          = false
	defaultGeneralLineBufferBytes    int64         = 16384
	defaultGeneralMaxLineBytes       int64         = 1048576
	defaultGeneralProspectInterval   time.Duration = 10 * time.Second
	defaultGeneralSpoolMaxBytes      int64         = 10485760
	defaultGeneralSpoolSize          int64         = 1024
	defaultGeneralTelemetryPath      string        = "/metrics"
	defaultGeneralSpoolTimeout       time.Duration = 5 * time.Second
	defaultNetworkBackoff            time.Duration = 5 * time.Second
	defaultNetworkBackoffMax         time.Duration = 300 * time.Second
	defaultNetworkMaxPendingPayloads int64         = 10
	defaultNetworkMethod             string        = "random"
	defaultNetworkRfc2782Service     string        = "courier"
	defaultNetworkRfc2782Srv         bool          = true
	defaultNetworkTimeout            time.Duration = 15 * time.Second
	defaultNetworkTransport          string        = "tls"
	defaultStreamAddHostField        bool          = true
	defaultStreamAddOffsetField      bool          = true
	defaultStreamAddPathField        bool          = true
	defaultStreamAddTimezoneField    bool          = false
	defaultStreamCodec               string        = "plain"
	defaultStreamDeadTime            time.Duration = 1 * time.Hour
	defaultStreamUnlinkedDeadTime    time.Duration = 1 * time.Hour
)

// Section is implemented by external config structures that will be
// registered with the config package
type Section interface {
	Validate() error
}

// SectionCreator creates new Section structures
type SectionCreator func() Section

// registeredSectionCreators contains a list of registered external Section
// creators that should be processed in all new Config structures
var registeredSectionCreators = make(map[string]SectionCreator)

// General holds the general configuration
type General struct {
	GlobalFields     map[string]interface{} `config:"global fields"`
	Host             string                 `config:"host"`
	LineBufferBytes  int64                  `config:"line buffer bytes"`
	ListenAddress    string                 `config:"listen address"`
	LogFile          string                 `config:"log file"`
	LogLevel         logging.Level          `config:"log level"`
	LogStdout        bool                   `config:"log stdout"`
	LogSyslog        bool                   `config:"log syslog"`
	MaxLineBytes     int64                  `config:"max line bytes"`
	PersistDir       string                 `config:"persist directory"`
	ProspectInterval time.Duration          `config:"prospect interval"`
	SpoolSize        int64                  `config:"spool size"`
	SpoolMaxBytes    int64                  `config:"spool max bytes"`
	SpoolTimeout     time.Duration          `config:"spool timeout"`
	TelemetryPath    string                 `config:"telemetry path"`
}

// InitDefaults initialises default values for the general configuration
func (gc *General) InitDefaults() {
	gc.LineBufferBytes = defaultGeneralLineBufferBytes
	gc.LogLevel = defaultGeneralLogLevel
	gc.LogStdout = defaultGeneralLogStdout
	gc.LogSyslog = defaultGeneralLogSyslog
	gc.MaxLineBytes = defaultGeneralMaxLineBytes
	gc.PersistDir = DefaultGeneralPersistDir
	gc.ProspectInterval = defaultGeneralProspectInterval
	gc.SpoolSize = defaultGeneralSpoolSize
	gc.SpoolMaxBytes = defaultGeneralSpoolMaxBytes
	gc.SpoolTimeout = defaultGeneralSpoolTimeout
	gc.TelemetryPath = defaultGeneralTelemetryPath
	// NOTE: Empty string for Host means calculate it automatically, so leave it
	// Likewise, an empty ListenAddress means to disable the Prometheus HTTP listener (the default)
}

// Network holds network related configuration
type Network struct {
	Factory      interface{}
	AddressPools []*addresspool.Pool

	Backoff            time.Duration `config:"failure backoff"`
	BackoffMax         time.Duration `config:"failure backoff max"`
	MaxPendingPayloads int64         `config:"max pending payloads"`
	Method             string        `config:"method"`
	Rfc2782Service     string        `config:"rfc 2782 service"`
	Rfc2782Srv         bool          `config:"rfc 2782 srv"`
	Servers            []string      `config:"servers"`
	Timeout            time.Duration `config:"timeout"`
	Transport          string        `config:"transport"`

	Unused map[string]interface{}
}

// InitDefaults initiases default values for the network configuration
func (nc *Network) InitDefaults() {
	nc.Backoff = defaultNetworkBackoff
	nc.BackoffMax = defaultNetworkBackoffMax
	nc.MaxPendingPayloads = defaultNetworkMaxPendingPayloads
	nc.Method = defaultNetworkMethod
	nc.Rfc2782Service = defaultNetworkRfc2782Service
	nc.Rfc2782Srv = defaultNetworkRfc2782Srv
	nc.Timeout = defaultNetworkTimeout
	nc.Transport = defaultNetworkTransport
}

// CodecStub holds an unknown codec configuration
// After initial parsing of configuration, these CodecStubs are turned into
// real configuration blocks for the codec given by their Name field
type CodecStub struct {
	Name    string `config:"name"`
	Unused  map[string]interface{}
	Factory interface{}
}

// Stream holds the configuration for a log stream
type Stream struct {
	AddHostField     bool                   `config:"add host field"`
	AddOffsetField   bool                   `config:"add offset field"`
	AddPathField     bool                   `config:"add path field"`
	AddTimezoneField bool                   `config:"add timezone field"`
	Codecs           []CodecStub            `config:"codecs"`
	DeadTime         time.Duration          `config:"dead time"`
	UnlinkedDeadTime time.Duration          `config:"unlinked dead time"`
	Fields           map[string]interface{} `config:"fields"`
}

// InitDefaults initialises the default configuration for a log stream
func (sc *Stream) InitDefaults() {
	sc.AddHostField = defaultStreamAddHostField
	sc.AddOffsetField = defaultStreamAddOffsetField
	sc.AddPathField = defaultStreamAddPathField
	sc.AddTimezoneField = defaultStreamAddTimezoneField
	sc.DeadTime = defaultStreamDeadTime
	sc.UnlinkedDeadTime = defaultStreamUnlinkedDeadTime
}

// File holds the configuration for a set of paths that share the same stream
// configuration
type File struct {
	Paths  []string `config:"paths"`
	Stream `config:",embed"`
}

// Config holds all the configuration for Log Courier
type Config struct {
	Files    []File   `config:"files"`
	General  General  `config:"general"`
	Includes []string `config:"includes"`
	Network  Network  `config:"network"`
	Stdin    Stream   `config:"stdin"`
	// Dynamic sections
	// TODO: All top level sections to use this
	Sections map[string]Section `config:",dynamic"`
}

// NewConfig creates a new, empty, configuration structure
func NewConfig() *Config {
	c := &Config{
		Sections: make(map[string]Section),
	}

	for name, creator := range registeredSectionCreators {
		c.Sections[name] = creator()
	}

	return c
}

// loadFile detects the extension of the given file and loads it using the
// relevant load function
func (c *Config) loadFile(filePath string, rawConfig interface{}) error {
	ext := path.Ext(filePath)

	switch ext {
	case ".json":
		return c.loadJSONFile(filePath, rawConfig)
	case ".conf":
		return c.loadJSONFile(filePath, rawConfig)
	case ".yaml":
		return c.loadYAMLFile(filePath, rawConfig)
	}

	return fmt.Errorf("File extension '%s' is not within the known extensions: conf, json, yaml", ext)
}

// Load the configuration from the given file
// If initFactories is false, factories (such as codec names or transport
// names) are not initialised so they do not need to be built in
func (c *Config) Load(path string, initFactories bool) (err error) {
	// Read the main config file
	rawConfig := make(map[string]interface{})
	if err = c.loadFile(path, &rawConfig); err != nil {
		return
	}

	// Populate configuration - reporting errors on spelling mistakes etc.
	if err = c.PopulateConfig(c, rawConfig, "/"); err != nil {
		return
	}

	// Iterate includes
	for _, glob := range c.Includes {
		// Glob the path
		var matches []string
		if matches, err = filepath.Glob(glob); err != nil {
			return
		}

		for _, include := range matches {
			// Read the include
			var rawInclude []interface{}
			if err = c.loadFile(include, &rawInclude); err != nil {
				return
			}

			// Append to configuration
			vRawInclude := reflect.ValueOf(rawInclude)
			if err = c.populateSlice(reflect.ValueOf(c).Elem().FieldByName("Files"), vRawInclude, fmt.Sprintf("%s/", include)); err != nil {
				return
			}
		}
	}

	if c.General.PersistDir == "" {
		err = fmt.Errorf("/general/persist directory must be specified")
		return
	}

	// Enforce maximum of 2 GB since event transmit length is uint32
	if c.General.SpoolMaxBytes > 2*1024*1024*1024 {
		err = fmt.Errorf("/general/spool max bytes can not be greater than 2 GiB")
		return
	}

	if c.General.LineBufferBytes < 1 {
		err = fmt.Errorf("/general/line buffer bytes must be greater than 1")
		return
	}

	// Max line bytes can not be larger than spool max bytes
	if c.General.MaxLineBytes > c.General.SpoolMaxBytes {
		err = fmt.Errorf("/general/max line bytes can not be greater than /general/spool max bytes")
		return
	}

	if c.General.Host == "" {
		ret, hostErr := os.Hostname()
		if hostErr == nil {
			c.General.Host = ret
		} else {
			c.General.Host = defaultGeneralHost
			log.Warning("Failed to determine the FQDN: %s", hostErr)
			log.Warning("Falling back to using default hostname: %s", c.General.Host)
		}
	}

	// Ensure all GlobalFields are map[string]interface{}
	if err = c.fixMapKeys("/general/global fields", c.General.GlobalFields); err != nil {
		return
	}

	// TODO: Network method factory in publisher
	if c.Network.Method == "" {
		c.Network.Method = defaultNetworkMethod
	}
	if c.Network.Method != "random" && c.Network.Method != "failover" && c.Network.Method != "loadbalance" {
		err = fmt.Errorf("The network method (/network/method) is not recognised: %s", c.Network.Method)
		return
	}

	if len(c.Network.Servers) == 0 {
		err = fmt.Errorf("No network servers were specified (/network/servers)")
		return
	}

	servers := make(map[string]bool)
	c.Network.AddressPools = make([]*addresspool.Pool, len(c.Network.Servers))
	for n, server := range c.Network.Servers {
		if _, exists := servers[server]; exists {
			err = fmt.Errorf("The list of network servers (/network/servers) must be unique: %s appears multiple times", server)
			return
		}
		servers[server] = true
		c.Network.AddressPools[n] = addresspool.NewPool(server)
	}

	if initFactories {
		if registrarFunc, ok := registeredTransports[c.Network.Transport]; ok {
			if c.Network.Factory, err = registrarFunc(c, "/network/", c.Network.Unused, c.Network.Transport); err != nil {
				return
			}
		} else {
			err = fmt.Errorf("Unrecognised transport '%s'", c.Network.Transport)
			return
		}
	}

	for k := range c.Files {
		if len(c.Files[k].Paths) == 0 {
			err = fmt.Errorf("No paths specified for /files[%d]/", k)
			return
		}

		if err = c.initStreamConfig(fmt.Sprintf("/files[%d]", k), &c.Files[k].Stream, initFactories); err != nil {
			return
		}
	}

	if err = c.initStreamConfig("/stdin", &c.Stdin, initFactories); err != nil {
		return
	}

	// Validate the registered configurables
	for _, section := range c.Sections {
		if err = section.Validate(); err != nil {
			return
		}
	}

	return
}

// initStreamConfig initialises a stream configuration by creating the necessary
// codec factories the harvesters will require
func (c *Config) initStreamConfig(path string, streamConfig *Stream, initFactories bool) (err error) {
	if !initFactories {
		// Currently only codec factory is initialised, so skip if we're not doing that
		return nil
	}

	if len(streamConfig.Codecs) == 0 {
		streamConfig.Codecs = []CodecStub{CodecStub{Name: defaultStreamCodec}}
	}

	for i := 0; i < len(streamConfig.Codecs); i++ {
		codec := &streamConfig.Codecs[i]
		if registrarFunc, ok := registeredCodecs[codec.Name]; ok {
			if codec.Factory, err = registrarFunc(c, path, codec.Unused, codec.Name); err != nil {
				return
			}
		} else {
			return fmt.Errorf("Unrecognised codec '%s' for %s", codec.Name, path)
		}
	}

	// Ensure all Fields are map[string]interface{}
	if err = c.fixMapKeys(path+"/fields", streamConfig.Fields); err != nil {
		return
	}

	// TODO: EDGE CASE: Event transmit length is uint32, if fields length is rediculous we will fail

	return nil
}

// Get returns the requested dynamic configuration entry
func (c *Config) Get(name string) interface{} {
	ret, ok := c.Sections[name]
	if !ok {
		return nil
	}

	return ret
}

// fixMapKeys converts any map entries where the keys are interface{} values
// into map entries where the key is a string. It returns an error if any key is
// found that is not a string.
// This is important as json.Encode will not encode a map where the keys are not
// concrete strings.
func (c *Config) fixMapKeys(path string, value map[string]interface{}) error {
	for k, v := range value {
		switch vt := v.(type) {
		case map[string]interface{}:
			if err := c.fixMapKeys(path+"/"+k, vt); err != nil {
				return err
			}
		case map[interface{}]interface{}:
			fixedValue, err := c.fixMapInterfaceKeys(path+"/"+k, vt)
			if err != nil {
				return err
			}

			value[k] = fixedValue
		}
	}

	return nil
}

func (c *Config) fixMapInterfaceKeys(path string, value map[interface{}]interface{}) (map[string]interface{}, error) {
	fixedMap := make(map[string]interface{})

	for k, v := range value {
		ks, ok := k.(string)
		if !ok {
			return nil, fmt.Errorf("Invalid non-string key at %s", path)
		}

		switch vt := v.(type) {
		case map[string]interface{}:
			if err := c.fixMapKeys(path+"/"+ks, vt); err != nil {
				return nil, err
			}

			fixedMap[ks] = vt
		case map[interface{}]interface{}:
			fixedValue, err := c.fixMapInterfaceKeys(path+"/"+ks, vt)
			if err != nil {
				return nil, err
			}

			fixedMap[ks] = fixedValue
		default:
			fixedMap[ks] = vt
		}
	}

	return fixedMap, nil
}

// RegisterConfigSection registers a new Section creator which will be used to
// create new sections that will be available via Get() in all created Config
// structures
func RegisterConfigSection(name string, creator SectionCreator) {
	registeredSectionCreators[name] = creator
}
